Add notes on block ordering and glRegDeps

a7ehuo opened the pull request:
This note captures the concepts of mainline code, out of line
code, and `GlRegDeps` that Vijay explained in
a meeting.

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

a7ehuo commented:
@vijaysun-omr Ready for review. Thanks!

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+-->
+
+# Mainline and out of line code
+The mainline path handles the common operations in a straight line to avoid branching or using branch prediction resources. Arranging the commonly run code on the mainline also helps reserve instruction cache. The work can be done in the same cache line by falling through. There is no need to jump to another spot to bring in an entire new cache line to the cache.
```
I think the "straight line" aspect is true usually but is not necessary, i.e. we could have some minor branching within the code on the mainline path to implement some basic logic on occasions. The main concept is that we feel there should be code that runs commonly in the mainline to take advantage of the branch and instruction cache resources without needing to branch to the out of line path that is usually in a different cache line.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+Executing the common code on the mainline is not only adopted in IL but also in the code gen, such as the non-trivial logics: write barrier, instanceof, etc. The instructions that are believed to run commonly are arranged on the mainline path. Branching is chosen for the less common code.
+
+One example in the code gen is the bound check. It does the comparison between arraylength and index. If it fails, an exception is going to be raised.  Then it will branch to an out of line, or a snippet of code that jumps into the VM to throw the exception that needs to be thrown. Because the path that jumps to throw the exception is expected to be very rarely called, the snippet code is put at  ...
+
```
The snippet to call into the VM to throw the appropriate exception could be at the end of the code section at the time of generating code for the bound check for example, but it is possible that subsequent bound checks (or other checks/operations)  that get evaluated also append more code and therefore the snippet for a given bound check is not at the very end but instead, near the end of the code section.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+One example in the code gen is the bound check. It does the comparison between arraylength and index. If it fails, an exception is going to be raised.  Then it will branch to an out of line, or a snippet of code that jumps into the VM to throw the exception that needs to be thrown. Because the path that jumps to throw the exception is expected to be very rarely called, the snippet code is put at  ...
+
+Another example is block ordering and extension of basic blocks in IL. Both uses the same logic to arrange the blocks. For example there is a block ending with `if`. One path is marked as cold. Another path is not marked as cold. Block ordering takes the cold path out of the mainline, puts it at the end of the method, and keeps the non-cold path as the fall through. This way it avoids consuming b ...
+
```
"For example..." sentence does not make it completely clear that the one path marked cold and the other not marked as cold is just one possible scenario.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+Most common control flow opcodes are: `if`, `goto`, `return`, `athrow`, `switch`. They should be the last tree before the `BBEnd`. They should have their own `GlRegDeps` for the branch target if required. The `GlRegDeps` under `BBEnd` is for the fall through case.
+
+If a value that is expected to be in a register is not in the register, or the register is not preserved if it is a call, it leads to register shuffling. The local register allocator has to shuffle it to a different register to make sure it has the right value in `GlRegDeps`. Register shuffling could come with a cost.
+
```
You probably mean "make sure it has the right value in the right real register at the point where the `GlRegDeps` is evaluated".

You probably mean "Register shuffling does come at a cost" since there is no doubt that there is extra instruction path length to do the shuffling.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+If a value that is expected to be in a register is not in the register, or the register is not preserved if it is a call, it leads to register shuffling. The local register allocator has to shuffle it to a different register to make sure it has the right value in `GlRegDeps`. Register shuffling could come with a cost.
+
+The local register allocator and the code gen use the reference count for `GlRegDeps` to determine when a register is no longer needed. It needs to be set up correctly, otherwise it could result in registers being overwritten with some other values prematurely.
+
```
The notion of using reference count for detecting when a register is no longer needed is not unique to `GlRegDeps` or other GRA related opcodes. i.e. it is just a basic rule within the codegen for nodes and their associated virtual registers in general.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+The local register allocator and the code gen use the reference count for `GlRegDeps` to determine when a register is no longer needed. It needs to be set up correctly, otherwise it could result in registers being overwritten with some other values prematurely.
+
+The `GlRegDeps` nodes should have been evaluated before they get to `GlRegDeps`. The `GlRegDeps` job is to take values that are already in virtual registers and simply map them to real registers. `GlRegDeps` is only about generating the register dependency. It should not be about generating any other instructions.
+
```
You probably mean something like "The children of the `GlRegDeps` node that represented new values assigned to new global registers in this extended block" when you say "The `GlRegDeps` nodes"

vijaysun-omr replied:
The only other instruction that a `GlRegDeps` evaluation can result in is a register-register move since we can have a situation where the same value (same IL node) goes out in two (or more) different global registers. In such a case we cannot have the same virtual register be mapped to two (or more) real registers in the codegen register dependency since this is a condition that cannot be satisfied. By adding a new virtual register and inserting a register-register copy instruction the codegen register dependency can use different virtual/real registers.

vijaysun-omr commented:
FYI @0xdaryl and @fjeremic since this is around the area of documentation. 

@a7ehuo should this be considered for the OMR project instead of OpenJ9 ? Another question is whether these topics ought to be split up in some way since they do touch somewhat different technical areas.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
````diff
@@ -0,0 +1,97 @@
+n190n           ==>iand
+n572n     BBEnd </block_33>                                                                   [0x7fec3b0ba280] bci=[1,30,-] rc=0 vc=0 vn=- li=- udi=- nc=0
+```
+
````
It is a little odd that the wrong and right examples above are not using the same basic set of IL trees. If I remember right you had a bug in the past couple of weeks where you had an evaluation of an arithmetic node (e.g. iadd) under the `GlRegDeps` and you fixed that later. This could be the example you use with the same basic set of IL trees showing the wrong and right way of doing the same thing.

a7ehuo replied:
I'll match the good example and the bad from the same method compilation in the next update.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+
+The `xRegStore` serves the purpose of anchoring those children that are honored and ensuring that they get evaluated before the `GlRegDeps` gets encountered. However, the `xRegStore` does not make sure that the value is in that real register at the point of `xRegStore`.
+
+`xRegStore` is not responsible for putting the value into the register. It is not actually going to do anything at that point other than evaluates the child and associates it with a virtual register. The actual moving off that value into the right real register is dictated by `GlRegDeps` at the end of block. The local register allocator in the code gen happens backwards. It walks backwards throug ...
```
It seems that the last sentence of the prior paragraph (starting with "However,") would be better as the first sentence of the next paragraph. Also, the sentence starting with "The local register allocator..." sounds like something that could be the start of the next paragraph too.

vijaysun-omr commented on doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md:
```diff
@@ -0,0 +1,97 @@
+`xRegStore` is not responsible for putting the value into the register. It is not actually going to do anything at that point other than evaluates the child and associates it with a virtual register. The actual moving off that value into the right real register is dictated by `GlRegDeps` at the end of block. The local register allocator in the code gen happens backwards. It walks backwards throug ...
+
+The purpose of `xRegStore` and `GlRegDeps` is only to specify which value should be in which real register at those points while transitioning from one block to another. It does not mandate the value has to be in this real register at this point. It ensures the right values are in the right real registers at those transition points.
+
```
The `xRegStore` does not mandate the value has to be in the right real register at that point but the `GlRegDeps` is in fact at the point where we transition from one block to another and so it is one of the points where the values have to be in the right registers.

a7ehuo commented:
> @a7ehuo should this be considered for the OMR project instead of OpenJ9 ? 

I think these nodes could be moved under` omr/doc/compiler/il/` https://github.com/eclipse/omr/tree/master/doc/compiler/il

> Another question is whether these topics ought to be split up in some way since they do touch somewhat different technical areas.

How about splitting it up into two: 
`Mainline_and_outOfLine.md`
`GlRegDeps.md`

a7ehuo commented:
I have moved the document to OMR: PR https://github.com/eclipse/omr/pull/5898. 
It's split into two documents: `GlRegDeps.md` and `MainlineAndOutOfLineCode.md`. All the above comments are also addressed in the OMR commit https://github.com/eclipse/omr/pull/5898/commits/5ef08b32cea97bdb7d233d98307494a7e81fe5d5.

a7ehuo commented:
This PR is closed. Please address all comments for this change in eclipse/omr#5898.
